src/features/manifest/validators.ts

```diff
@@ -0,0 +1,98 @@
+import { SnapManifest, SnapManifestStruct } from '@metamask/snaps-utils';
+
+type ValidatorContext = {
+  sourceCode: string;
+  icon: string;
+};
+
+type ValidatorFunction = (
+  value: SnapManifest,
+  context: ValidatorContext,
+) => Promise<boolean | string>;
+
+export type Validator = {
+  name: string;
+  manifestName: string;
+  validator: ValidatorFunction;
+};
+
+export const validators: Validator[] = [
+  {
+    name: 'Title',
+    manifestName: 'proposedName',
+    validator: async (manifest: SnapManifest) => {
+      const [error] = SnapManifestStruct.schema.proposedName.validate(
+        manifest?.proposedName,
+      );
+      return error?.message ?? true;
+    },
+  },
+  {
+    name: 'Description',
+    manifestName: 'description',
+    validator: async (manifest: SnapManifest) => {
+      const [error] = SnapManifestStruct.schema.description.validate(
+        manifest?.description,
+      );
+      return error?.message ?? true;
+    },
+  },
+  {
+    name: 'Icon',
+    manifestName: 'iconPath',
+    validator: async (manifest: SnapManifest, { icon }) => {
+      const [error] =
+        SnapManifestStruct.schema.source.schema.location.schema.npm.schema.iconPath.validate(
+          manifest?.source?.location?.npm?.iconPath,
+        );
+
+      if (error) {
+        return error.message;
+      }
+
+      if (!icon) {
+        return 'Unable to load icon.';
+      }
+
+      return true;
+    },
+  },
+  {
+    name: 'Permissions',
+    manifestName: 'initialPermissions',
+    validator: async (manifest: SnapManifest) => {
+      const [error] = SnapManifestStruct.schema.initialPermissions.validate(
```

Comment:
Should we also process the permissions and validate them with the PermissionController?